core/actionProxy/actionproxy.py

```diff
@@ -139,9 +142,9 @@ def error(msg):
 
         # stdout/stderr may be either text or bytes, depending on Python
         # version.   In the latter case, decode to text.
-        if isinstance(o, bytes):
+        if not isinstance(o, str):
```

Comment:
Does this change work in both Python 2 and 3 environments?  (This is really a side question: Does OpenWhisk run the Docker action tests using both Python 2 and Python 3?)

In the end, we don't care about the version of Python (and [should avoid using version detection](https://docs.python.org/3/howto/pyporting.html#use-feature-detection-instead-of-version-detection)); we want `actionProxy.py` to handle the output from actions robustly.  

#1867 was originally opened to handle `Popen` returning text in Python 2 and bytes in Python 3.  I'm assuming the tests you mention are failing because of a unicode error?  In other words, this is not a bytes vs. text issue, but `actionProxy.py` needs to handle output that contains unicode characters?

I keep seeing the same advice when dealing with text in Python 2/3, which is to convert text at the edge. https://pythonhosted.org/kitchen/unicode-frustrations.html#a-few-solutions

> If you get some piece of text from a library, read from a file, etc, turn it into a unicode string immediately. 

This fix probably works in most cases, but there's a lot of implicit logic in it:

```
if not isinstance(o, str): 
    o = o.decode('utf-8')
```

The original `if isinstance(o, bytes):` was really intended to say, "are you bytes from Python 3? if so, do the extra step to convert to unicode str; otherwise, skip this step".

But it seems we need to decode regardless; that is, convert `str` to `unicode` (Python 2), or `bytes` to `str`, which is really unicode (Python 3).

To recap: 

Python 3: as written, subprocess output is always `bytes`, use `decode` to convert to unicode `str`

```
Python 3.5.2
>>> import sys, subprocess
>>> p = subprocess.Popen(['echo', '{"foo":"и"}'], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
>>> o, e = p.communicate()
>>> type(o)
<class 'bytes'>
>>> o = o.decode('utf-8')
>>> type(o)
<class 'str'>
>>> process_output_lines = o.strip().split('\n')
>>> sys.stdout.write(o)
{"foo":"и"}
```

Python 2: as written, subprocess output is always `str`, use `decode` to convert to `unicode` (same as `str` in Python 3)

```
Python 2.7.13 
>>> import sys, subprocess
>>> p = subprocess.Popen(['echo', '{"foo":"и"}'], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
>>> o, e = p.communicate()
>>> type(o)
<type 'str'>
>>> o = o.decode('utf-8')
>>> type(o)
<type 'unicode'>
>>> process_output_lines = o.strip().split('\n')
>>> sys.stdout.write(o)
{"foo":"и"}
``` 

So maybe the correct way to handle is to `decode` always?  If so, I wonder if we should skip the manual step and instead force `Popen` to give us text always.  The `encoding` parameter was added in Python 3, but it does not exist in Python 2.

https://docs.python.org/3/library/subprocess.html#frequently-used-arguments

> If encoding or errors are specified, or universal_newlines is true, the file objects stdin, stdout and stderr will be opened in text mode using the encoding and errors specified in the call or the defaults for io.TextIOWrapper.

The only way I see to force text output in both Python 2 and 3 is to use `universal_newlines=True`, which seems hacky.

In any case, maybe more comments in the code will help folks understand what's going on.